### PR TITLE
Complete the Jest DI with crypto

### PR DESCRIPTION
**Pull request by matheus23**
This is preparation work for making it possible to write more sophisticated test, making test-driven-development for internal (non-browser-bundle-exposed) APIs possible.

### CODE REVIEW

File: src/common/version.ts [hunk 1]
```diff
@@ -1 +1 @@
-export const VERSION = "0.24.0"
+export const VERSION = "0.24.2"
```
Comment (matheus23) [inline src/common/version.ts:1]: No idea why this was stale :/

Maybe we should .gitignore this?

File: src/setup/jest.ts [hunk 2]
```diff
@@ -10,40 +15,306 @@ import { setDependencies } from './dependencies'
...
+  async getSymmKey(keyName: string, cfg?: Partial<Config>): Promise<CryptoKey> {
+    const mergedCfg = config.merge(this.cfg, cfg)
+    const maybeKey = this.store[keyName]
+    if(maybeKey !== null) {
+      return maybeKey
+    }
+    const key = await aes.makeKey(config.symmKeyOpts(mergedCfg))
+    this.store[keyName] = key
+    return key
+  }
```
Comment (matheus23) [inline src/setup/jest.ts:128]: I duplicate some code here [from keystore-idb](https://github.com/fission-suite/keystore-idb/blob/74d2a96429dde2b86d3daa3597b4ef35334db937/src/rsa/keystore.ts).

We should probably eventually make it possible to use our dependencies in node. However, doing that for keystore-idb could be questionable, given it has "idb" in its name :smile: 

Any way, this feels kind of quick-and-dirty right now. Let me know if you have any other ideas.

File: src/index.ts [hunk 3]
```diff
@@ -256,6 +256,7 @@ export * as ipfs from './ipfs'
 export * as keystore from './keystore'
 export * as machinery from './common'
 export * as crypto from './crypto'
+export * as cbor from 'cborg'
```
Comment (matheus23) [inline src/index.ts:259]: I'm not happy about exposing this, but I need this to test the browser build in `encoding.test.ts`

File: package.json [hunk 4]
```diff
@@ -36,7 +36,7 @@
     "prebuild": "rimraf dist && node scripts/gen-version.js",
     "build": "tsc && rollup -c rollup.config.ts",
     "start": "tsc -w",
-    "test": "jest",
+    "test": "jest --forceExit",
```
Comment (walkah) [inline package.json:39]: Hrm. This smells like there's something wrong with the tests... async operations not resolving?

File: src/fs/protocol/private/mmpt.test.ts [hunk 5]
```diff
@@ -1,8 +1,9 @@
 import crypto from "crypto"
-import Ipfs, { IPFS } from "ipfs"
+import { IPFS } from "ipfs-core"
```
Comment (walkah) [inline src/fs/protocol/private/mmpt.test.ts:2]: Is this a size thing? (full IPFS includes more code than we use, etc?)

**Review by walkah — APPROVED** [review]
This is great! thanks for your continued efforts getting this working!

I Appreciate You - KD22dF8apIKLSapKSC

FYI: `bcrypto` isn't currently building for me locally in the nix shell on macOS. I'm going to poke around, but it seems to work in CI so may be my environment.

File: src/fs/protocol/private/mmpt.test.ts (hunk 5 above)
Comment (matheus23) [reply src/fs/protocol/private/mmpt.test.ts:2]: Yep, exactly. It's just much less dependencies.
I think the difference is that ipfs also contains the CLI and an http-server, which ipfs-core does not.

File: package.json (hunk 4 above)
Comment (matheus23) [reply package.json:39]: Yep. I tried to track it down for a while, but gave up. From my testing it seems like in some edge-cases it's possible that ipfs leaves a child process and some sockets hanging after `await ipfs.stop()`. Not sure what's up with that, but after fighting with it for a while I just gave up.
The IPFS we're using in the tests is basically in-memory only. It doesn't write to disk. So killing it shouldn't be an issue.

Comment (walkah) [reply package.json:39]: Cool. yeah, something we can track down and fix as a follow up later, imo.

**Comment by matheus23** [comment]
> FYI: `bcrypto` isn't currently building for me locally in the nix shell on macOS. I'm going to poke around, but it seems to work in CI so may be my environment.

Yeah. Weird. `bcrypto` is building fine on my end (including in the nix shell).

**Review by bgins — APPROVED** [review]
Looks great! Thanks for implementing these Jest DI crypto functions :tada: :pray:

File: tests/encoding.test.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        expect(decoded).toEqual(message)
+    })
+
+    it("works in the browser with encryption in between", async () => {
+        await loadWebnativePage()
+
+        async function runRoundTrip(message) {
+            const keyStr = await webnative.crypto.aes.genKeyStr()
+
+            const encoded = webnative.cbor.encode(message)
```
[the remark this answers is not in the document]
Comment (matheus23) [reply tests/encoding.test.ts:30]: Did you `yarn build` before you `yarn test`ed?

[the remark this answers is not in the document]
Comment (bgins) [reply tests/encoding.test.ts:30]: I did not :facepalm: and that was it. After building, this test and all the others are passing for me :100:

[the remark this answers is not in the document]
Comment (matheus23) [reply tests/encoding.test.ts:30]: Admittedly, it's really confusing. The puppeteer html file fixture uses the index.umd.min.js file, so unless one `yarn build`s before running the tests, it'll use an outdated webnative, but only for the puppeteer stuff.

**Pull request merged**